### ISSUE TITLE
Type-check macro arguments manually

### ISSUE THREAD

**Issue by Profpatsch** [tweag/rules_haskell#1211]
**Describe the bug**

We don’t check the types of arguments passed to macros we define, so users can get very confusing error messages when they update.

For example, the `extra_deps` field to `stack_snapshot` changed to be an attrset instead of a list, leading to the following error when updating:

```
ERROR: /home/philip/.cache/bazel/_bazel_philip/ec4bffc753a4896ad5f0595f4cbba0c3/external/rules_haskell/haskell/cabal.bzl:725:37: Traceback (most recent call last):
	File "/home/philip/kot/work/tweag/zeuslogics/WORKSPACE", line 108
		stack_snapshot(<5 more arguments>)
	File "/home/philip/.cache/bazel/_bazel_philip/ec4bffc753a4896ad5f0595f4cbba0c3/external/rules_haskell/haskell/cabal.bzl", line 1036, in stack_snapshot
		_stack_snapshot(stack = stack, <3 more arguments>)
	File "/home/philip/.cache/bazel/_bazel_philip/ec4bffc753a4896ad5f0595f4cbba0c3/external/rules_haskell/haskell/cabal.bzl", line 1040, in _stack_snapshot
		_from_string_keyed_label_list_dict(<1 more arguments>)
	File "/home/philip/.cache/bazel/_bazel_philip/ec4bffc753a4896ad5f0595f4cbba0c3/external/rules_haskell/haskell/cabal.bzl", line 725, in _from_string_keyed_label_list_dict
		d.items()
type 'list' has no method items()
ERROR: error loading package 'external': Package 'external' contains errors
```

This can be circumvented if we defensively type-check our macro’s inputs.

**To Reproduce**

Write a `WORKSPACE` file like

[WORKSPACE.txt](https://github.com/tweag/rules_haskell/files/4054012/WORKSPACE.txt)

and run `bazel build @stackage://...`

**Expected behavior**

The error thrown should be something like:

“stackage_snapshot `extra_deps` requires a dict from dependency name to list of targets, but was given a list of targets” 

**Environment**
 - OS name + version: nixos
 - Bazel version: 2.0.0
 - Version of the rules: head (see attached WORKSPACE file)

**Comment by smelc** [comment]
Do you have a particular set of rules for which you'd like to have better error messages thanks to typing tests?

For example, we could add tests for the rules for which there are examples (output shows the number of `Examples` section within each file):

```
> grep '### Examples' haskell/*.bzl -c | grep -v '0$'
haskell/c2hs.bzl:1
haskell/cabal.bzl:3
haskell/defs.bzl:3
haskell/doctest.bzl:1
haskell/ghc_bindist.bzl:1
haskell/haddock.bzl:1
haskell/nixpkgs.bzl:1
haskell/protobuf.bzl:2
haskell/repl.bzl:1
haskell/toolchain.bzl:1
```

I suppose these are the most important rules.

**Comment by Profpatsch** [comment]
The attributes rules proper already have types that are verified by bazel. Only the macros (that is `def`s that are calling rules like functions) don’t get type-checked, yet we heavily use them, for example in `stack_snapshot`.

**Comment by mboes** [comment]
And even then, most macros don't do anything in particular to the arguments, beyond passing them to rules. So I think we should address the particular case in the description of this ticket, close this ticket, and then address similar issues elsewhere as they arise.

**Comment by smelc** [comment]
Closed via #1270
